python-stdlib/logging/logging.py

```diff
@@ -1,128 +1,212 @@
 
-    def __init__(self, name):
+    def emit(self, record):
+        if record.levelno >= self.level:
+            self.stream.write(self.format(record) + self.terminator)
+
+
+class FileHandler(StreamHandler):
+    def __init__(self, filename, mode="a", encoding="UTF-8"):
+        super().__init__(stream=open(filename, mode=mode, encoding=encoding))
+
+    def close(self):
+        super().close()
+        self.stream.close()
+
+
+class LogRecord:
+    def set(self, name, level, message):
         self.name = name
+        self.levelno = level
+        self.levelname = _level_dict[level]
+        self.message = message
+        self.ct = time.time()
+        self.msecs = int((self.ct - int(self.ct)) * 1000)
+        self.asctime = None
+
+
+class Formatter:
+    def __init__(self, fmt=default_fmt, datefmt=default_datefmt):
+        self.fmt = fmt
+        self.datefmt = datefmt
+
+    def usesTime(self):
+        return "asctime" in self.fmt
+
+    def formatTime(self, datefmt, record):
+        if hasattr(time, "strftime"):
+            return time.strftime(datefmt, time.localtime(record.ct))
+        return ""
+
+    def format(self, record):
+        if self.usesTime():
+            record.asctime = self.formatTime(self.datefmt, record)
+        return self.fmt % {
+            "name": record.name,
+            "message": record.message,
+            "msecs": record.msecs,
+            "asctime": record.asctime,
+            "levelname": record.levelname,
+        }
 
-    def _level_str(self, level):
-        l = _level_dict.get(level)
-        if l is not None:
-            return l
-        return "LVL%s" % level
+
+class Logger:
+    def __init__(self, name):
+        self.name = name
+        self.level = NOTSET
+        self.handlers = []
+        self.record = LogRecord()
 
     def setLevel(self, level):
         self.level = level
 
+    def addHandler(self, handler):
+        self.handlers.append(handler)
+
+    def hasHandlers(self):
+        return len(self.handlers) > 0
+
     def isEnabledFor(self, level):
-        return level >= (self.level or _level)
+        return level >= self.level
 
-    def log(self, level, msg, *args):
-        if self.isEnabledFor(level):
-            levelname = self._level_str(level)
-            if args:
-                msg = msg % args
-            if self.handlers:
-                d = self.record.__dict__
-                d["levelname"] = levelname
-                d["levelno"] = level
-                d["message"] = msg
-                d["name"] = self.name
-                for h in self.handlers:
-                    h.emit(self.record)
-            else:
-                print(levelname, ":", self.name, ":", msg, sep="", file=_stream)
+    def debug(self, message, *args):
+        self.log(DEBUG, message, *args)
 
-    def debug(self, msg, *args):
-        self.log(DEBUG, msg, *args)
+    def info(self, message, *args):
+        self.log(INFO, message, *args)
```

Comment:
Please keep the original order of these methods, and also the original argument names.  There's no need to change this.